lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java

```diff
@@ -639,13 +639,15 @@ public TopFieldDocs reduce(Collection<TopFieldCollector> collectors) throws IOEx
       for (int i = 0; i < leafSlices.length; ++i) {
         final LeafReaderContext[] leaves = leafSlices[i].leaves;
         final C collector = collectors.get(i);
-        topDocsFutures.add(executor.submit(new Callable<C>() {
-          @Override
-          public C call() throws Exception {
+        if (i == leafSlices.length - 1) { // execute the last on the caller thread
```

Comment:
Can we change `for` loop to only iterate up to `leafSlices.length - 1` and then do this afterwards?

Reply:
Yeah, would be better!

Reply:
I pushed new changes